src/Utilities.hs

```diff
@@ -0,0 +1,85 @@
+module Utilities (noInlineTypecheck, compileToCore) where
+
+import Data.Data (Data, gmapT)
+import Data.Generics (extT)
+import GHC
+import GHC.Driver.Main
+import GHC.Paths (libdir)
+import GHC.Plugins
+import GHC.Tc.Types
+import System.FilePath (takeBaseName)
+
+-- | Custom `compileToCore` function which compiles a haskell module at a
+-- specified file path into GHC Core. Returns a `CoreProgram` and the internally
+-- defined and used `DynFlags`. This flags are used for safe pretty printing.
+compileToCore :: FilePath -> IO (CoreProgram, DynFlags)
+compileToCore filePath = runGhc (Just libdir) $ do
+  dflags <- getSessionDynFlags
+  let newDflags =
+        dflags
+          { ghcLink = NoLink,
+            ghcMode = CompManager,
+            backend = interpreterBackend,
+            verbosity = 0,
+            debugLevel = 0
+          }
+  _ <- setSessionDynFlags newDflags
+  target <- guessTarget filePath Nothing Nothing
+  setTargets [target]
+  _ <- load LoadAllTargets
+  modSummary <- getModSummary $ mkModuleName (takeBaseName filePath)
+  env <- getSession
+  parsedModule <- liftIO $ hscParse env modSummary
+  (tcg, _) <- liftIO $ hscTypecheckRename env modSummary parsedModule
+  let noInlineTcg = noInlineTypecheck tcg
```

Comment:
Oh! So this just applies as an intermediate step in the compilation process, no need for template haskell. Nice!